File: openhoof/tool_registry.py
"""
Tools — Tool registry and execution wrapper.

Compatible with OpenHoof tool schemas:
    {
        "name": "drone_takeoff",
        "description": "Take off and hover at specified altitude",
        "parameters": {
            "type": "object",
            "properties": {
                "alt_m": {"type": "number", "default": 15.0}
            }
        }
    }

The ToolRegistry:
1. Registers tools from OpenHoof-compatible schema
2. Routes tool calls by name
3. Validates parameters (basic)
4. Captures training data for every call
"""

from __future__ import annotations

import json
import time
from typing import Any, Callable, Dict, List, Optional


class Tool:
    """A single tool definition."""
    
    def __init__(self, schema: Dict[str, Any], executor: Optional[Callable] = None):
        self.name = schema["name"]
        self.description = schema.get("description", "")
        self.parameters = schema.get("parameters", {})
        self.executor = executor
    
    def execute(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute the tool with given parameters."""
        if self.executor is None:
            return {
                "status": "error",
                "error": f"No executor registered for tool {self.name!r}"
            }
        
        try:
            result = self.executor(self.name, params)
            return result if isinstance(result, dict) else {"result": result}
        except Exception as e:
            return {
                "status": "error",
                "error": str(e),
                "tool": self.name
            }
    
    def __repr__(self) -> str:
        return f"Tool(name={self.name!r}, has_executor={self.executor is not None})"


class ToolRegistry:
    """Registry of available tools."""
    
    def __init__(self, tools: List[Dict[str, Any]] = None, executor: Optional[Callable] = None):
        """
        Initialize tool registry.
        
        Args:
            tools: List of OpenHoof-compatible tool schemas
            executor: Callable that executes tools: executor(tool_name, params) -> result
        """
        self.tools: Dict[str, Tool] = {}
        self.executor = executor
        self.call_count = 0
        
        if tools:
            self.register_tools(tools, executor)
    
    def register_tools(self, tools: List[Dict[str, Any]], executor: Optional[Callable] = None):
        """Register multiple tools at once."""
        exec_fn = executor or self.executor
        for schema in tools:
            # Support both OpenAI format and direct format
            if "function" in schema:
                # OpenAI format: {"type": "function", "function": {...}}
                tool_schema = schema["function"]
            else:
                # Direct format: {"name": "...", "description": "...", ...}
                tool_schema = schema
            
            name = tool_schema["name"]
            self.tools[name] = Tool(tool_schema, exec_fn)
    
    def register(self, schema: Dict[str, Any], executor: Optional[Callable] = None):
        """Register a single tool."""
        name = schema["name"]
        exec_fn = executor or self.executor
        self.tools[name] = Tool(schema, exec_fn)
    
    def get(self, name: str) -> Optional[Tool]:
        """Get a tool by name."""
        return self.tools.get(name)
    
    def execute(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a tool by name."""
        self.call_count += 1
        
        tool = self.get(tool_name)
        if tool is None:
            return {
                "status": "error",
                "error": f"Tool {tool_name!r} not found. Available: {list(self.tools.keys())}"
            }
        
        return tool.execute(params)
    
    def list_tools(self) -> List[str]:
        """List all registered tool names."""
        return list(self.tools.keys())
    
    def to_schema(self, format: str = "openai") -> List[Dict[str, Any]]:
        """
        Export all tools as schemas.
        
        Args:
            format: "openai" (with type/function wrapper) or "simple" (direct)
        
        Returns:
            List of tool schemas
        """
        if format == "openai":
            return [
                {
                    "type": "function",
                    "function": {
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": tool.parameters
                    }
                }
                for tool in self.tools.values()
            ]
        else:
            # Simple format (backward compatible)
            return [
                {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters
                }
                for tool in self.tools.values()
            ]
    
    def __len__(self) -> int:
        return len(self.tools)
    
    def __repr__(self) -> str:
        return f"ToolRegistry(tools={len(self.tools)}, calls={self.call_count})"

File: kotlin/src/main/kotlin/ai/llamafarm/openhoof/ToolRegistry.kt
package ai.llamafarm.openhoof

import kotlinx.serialization.json.*

/**
 * ToolRegistry — registers tools and dispatches calls.
 *
 * A tool is:
 *   - A schema (OpenAI function-calling format JSON)
 *   - A handler (suspend function that takes params, returns Any?)
 *
 * Supports:
 *   - Manual registration (register { ... })
 *   - Bulk registration from JSON schema list
 *   - OpenHoof Python-compatible schema format
 *
 * Usage:
 *   val registry = ToolRegistry()
 *
 *   registry.register(
 *     name = "get_wan_status",
 *     description = "Get WAN connection status",
 *     parameters = mapOf("type" to "object", "properties" to mapOf<String, Any>()),
 *   ) { _ ->
 *     mapOf("status" to "connected", "ip" to "1.2.3.4")
 *   }
 *
 *   val result = registry.execute("get_wan_status", emptyMap())
 */
class ToolRegistry {

    private val tools = mutableMapOf<String, RegisteredTool>()

    // ── Registration ──────────────────────────────────────────────────────────

    /** Register a tool with explicit schema components. */
    fun register(
        name: String,
        description: String = "",
        parameters: Map<String, Any> = emptyMap(),
        handler: suspend (params: Map<String, Any>) -> Any?,
    ) {
        val schema = buildJsonObject {
            put("type", "function")
            putJsonObject("function") {
                put("name", name)
                if (description.isNotBlank()) put("description", description)
                put("parameters", mapToJson(parameters))
            }
        }
        tools[name] = RegisteredTool(name = name, schema = schema, handler = handler)
    }

    /** Register a tool directly from an OpenAI-format JSON schema + handler. */
    fun register(schema: JsonObject, handler: suspend (params: Map<String, Any>) -> Any?) {
        val fn = schema["function"]?.jsonObject ?: schema
        val name = fn["name"]?.jsonPrimitive?.content
            ?: throw IllegalArgumentException("Tool schema missing 'name'")
        val normalized = if (schema.containsKey("function")) schema else buildJsonObject {
            put("type", "function")
            put("function", schema)
        }
        tools[name] = RegisteredTool(name = name, schema = normalized, handler = handler)
    }

    /** Register multiple tools from a JSON array of OpenAI-format schemas. */
    fun registerAll(schemas: JsonArray, handler: (name: String) -> (suspend (Map<String, Any>) -> Any?)) {
        schemas.forEach { element ->
            val obj = element.jsonObject
            val fn = obj["function"]?.jsonObject ?: obj
            val name = fn["name"]?.jsonPrimitive?.content ?: return@forEach
            register(obj) { params -> handler(name)(params) }
        }
    }

    // ── Execution ─────────────────────────────────────────────────────────────

    /** Execute a tool call. Throws [UnknownToolException] if not registered. */
    suspend fun execute(name: String, params: Map<String, Any>): Any? {
        val tool = tools[name] ?: throw UnknownToolException(name, tools.keys.toList())
        return tool.handler(params)
    }

    /** Execute, returning a Result wrapper instead of throwing. */
    suspend fun executeSafe(name: String, params: Map<String, Any>): ToolResult {
        return try {
            ToolResult.Success(name, execute(name, params))
        } catch (e: UnknownToolException) {
            ToolResult.UnknownTool(name, tools.keys.toList())
        } catch (e: Exception) {
            ToolResult.Error(name, e.message ?: e.toString())
        }
    }

    // ── Schema access ─────────────────────────────────────────────────────────

    /** All registered tool schemas as a JsonArray (for passing to ModelClient). */
    val schemas: JsonArray get() = buildJsonArray {
        tools.values.forEach { add(it.schema) }
    }

    /** Names of all registered tools. */
    val names: Set<String> get() = tools.keys.toSet()

    fun contains(name: String): Boolean = name in tools

    val size: Int get() = tools.size

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun mapToJson(map: Map<String, Any>): JsonElement = when (map["type"]) {
        "object" -> buildJsonObject {
            put("type", "object")
            @Suppress("UNCHECKED_CAST")
            (map["properties"] as? Map<String, Any>)?.let { props ->
                putJsonObject("properties") {
                    props.forEach { (k, v) ->
                        put(k, if (v is Map<*, *>) {
                            @Suppress("UNCHECKED_CAST")
                            mapToJson(v as Map<String, Any>)
                        } else JsonPrimitive(v.toString()))
                    }
                }
            }
            (map["required"] as? List<*>)?.let { req ->
                putJsonArray("required") { req.forEach { add(it.toString()) } }
            }
        }
        else -> buildJsonObject {
            map.forEach { (k, v) ->
                when (v) {
                    is String  -> put(k, v)
                    is Number  -> put(k, v.toDouble())
                    is Boolean -> put(k, v)
                    is List<*> -> putJsonArray(k) { v.forEach { add(it.toString()) } }
                    else       -> put(k, v.toString())
                }
            }
        }
    }
}

// ── Supporting types ──────────────────────────────────────────────────────────

private data class RegisteredTool(
    val name: String,
    val schema: JsonObject,
    val handler: suspend (Map<String, Any>) -> Any?,
)

sealed class ToolResult {
    abstract val toolName: String

    data class Success(override val toolName: String, val value: Any?) : ToolResult()
    data class Error(override val toolName: String, val message: String) : ToolResult()
    data class UnknownTool(override val toolName: String, val available: List<String>) : ToolResult()
}

class UnknownToolException(name: String, available: List<String>) :
    Exception("Unknown tool '$name'. Available: ${available.joinToString()}")
